ciao-deploy/cmd/join.go

```diff
@@ -30,10 +31,15 @@ func join(args []string) int {
 	ctx, cancelFunc := getSignalContext()
 	defer cancelFunc()
 
+	var debugLog bytes.Buffer
+
+	ctx = deploy.WithDebugLogWriter(ctx, &debugLog)
```

Comment:
bytes.Buffer is not thread-safe.  The fact that you're passing it to a context suggests that the buffer might be used by different go routines concurrently.  Is this the case?  If so, I think you're going to need some locking.